text/0022-pibd-deployment.md

```diff
@@ -0,0 +1,444 @@
+
+```
+Height
+1         2      5  
+         / \    / \ 
+0       0   1 (3) (4)
+
+Size: 6
+Last Position: 5
+```
+
+Then calculate hashes up to the position of `next leaf index - 1` (i.e. 7 - 1):
+
+```
+Height
+2             6
+           /     \
+1         2       5  
+         / \     / \ 
+0       0   1  (3) (4)
+
+Size: 7
+Last Position: 6
+```
+
+The next segment element will be the hash at 6, we check this matches our calculated hash and continue.
+
+At this point, the next elements could be (for example):
+
+* Leaves at 7 and 8 - in which case we'd simply apply the leaves and calculate hashes to the next leaf insertion index, giving:
+
+```
+Height
+2             6       
+           /     \    
+1         2       5       9 
+         / \     / \     / \ 
+0       0   1  (3) (4)  7   8
+
+Size: 10
+Last Position: 9
+```
+
+* A hash for 9 (if leaves 7 and 8 are pruned), in which case we'd apply the pruned subtree and calculate hashes to the next insertion index, giving:
+
+```
+Height
+2             6       
+           /     \    
+1         2       5       9 
+         / \     / \     / \ 
+0       0   1  (3) (4) (7) (8)
+
+Size: 10
+Last Position: 9
+```
+
+* A hash at 13 (if leaves 7, 8, 10 and 11 are pruned), in which case we apply the pruned subtree and calculate hashes to the next insertion index, giving:
+
+```
+Height
+3                     14 
+                /             \
+               /               \
+2             6                 13   
+           /     \          /        \
+1         2       5       (9)       (12)
+         / \     / \      / \      /    \ 
+0       0   1  (3) (4)  (7) (8)  (10)  (11)
+
+Size: 14
+Last Position: 13
+```
+
+This process continues until all hash and leaf values in the segment have been exhausted.
+
+If there is a mismatch anywhere or an error resulting from applying an element to an MMR, the node should wipe all data and restart the PIBD process. (Note this would most likely be the result of an internal or implementation error.)
+
+#### Determining MMR Completion
+
+An MMR can be considered 'complete' if its size matches the corresponding size and MMR root values in the horizon header.
+
+The entire PIBD process can be considered complete if all 3 MMRs are complete, and the horizon header is up-to-date.
+
+#### Post PIBD validation
+
+After validating and applying all segments, the node must verify all transaction objects as per existing fast-sync requirements. This includes.
+
+* Checking all MMR roots
+* Validating Kernels
+* Validating Rangeproofs
+
+Note that before validating transaction objects, the node must perform a full pass through the local leaf data of the output and rangeproof MMRs to ensure any outputs that may have been newly-spent are marked as such (or MMR root validation is likely to fail). This is required particularly when catching-up after a long period of the node being offline, or as a the result of the horizon header changing during the PIBD process.
```

Comment:
What is meant by "before validating transaction objects" ? When all MMR roots at the horizon height are verified, and kernels and rangeproofs at the horizon are verified, and sync is ready to apply post-horizon blocks?

Reply:
Clarified in document